### ISSUE TITLE
Fetch parachain runtime code from the relay chain

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot-sdk#78]
Currently the runtime code of the parachain is always fetched from the parachain state. This is done automatically by Substrate by fetching the `:code` storage item.
For parachains however, there actually exist two places the runtime code can be found, in the parachain state or in the relay chain state.

When there is a problem with the parachain, e.g. it is stalled and it needs to be recovered, it could happen that we need to replace the last head and the runtime on the relay chain. Then it would be nice to have all parachain nodes fetch the correct runtime from the relay chain to have this fixed, instead of also needing to coordinate with them to replace the runtime.

TLDR: We need to change Substrate to support fetching the runtime code from the relay chain, instead of the parachain state.

**Comment by xlc** [comment]
One thing to consider is for node to perform a full sync. It may not be able to get the wasm from relaychain once it is no longer the current one. Such wasm may not exist on parachain state at all.

I guess the solution is to be able to set such wasm in chainspec, similar to `code_substitutes`.

**Comment by burdges** [comment]
I suppose archive nodes could keep all the old code too since they keeps the blocks that provided it?

**Comment by kianenigma** [comment]
Can the `:code` still be part of the parachain state, but on every block, what is set in the relay chain can automatically replace it (within the parachain inherent thingy that passes a number of relay state proofs to the parachain runtime)?

This will fully remove the ability of the parachain to upgrade its code from within its own state. A parachain runtime upgrade should upgrade the code stored in the relay chain. 

Not sure if the above is feasible, but it will:

1. ensure that the historical codes are stored in the parachain state
2. there is only one source of truth

**Comment by bkchr** [comment]
> A parachain runtime upgrade should upgrade the code stored in the relay chain.

It already does this today.

**Comment by burdges** [comment]
A collator tracks both the relay chain and the parachain, so afaik the real question here is:  how do parachain archive nodes get the code?  We've never much discussed how para archive nodes track the relay chain and grandpa, maybe they archive the relay chain too, which maybe overkill.  We'd need the para archive node archive at least those relay chain blocks that contain their code upgrades.   Or maybe have some thin archive node notion which ignores the code?

**Comment by bkchr** [comment]
Every parachain node is tracking the relay chain and the parachain, because it gets finality etc from the relay chain. I would keep the code in the parachain state as well, this just makes it easier. When the code changes on the relay chain, the collator could input the new code using a relay chain proof to the parachain. This way the parachain state also gets updated.
